py/runtime.c

```diff
@@ -1409,3 +1409,76 @@ NORETURN void mp_raise_TypeError(const char *msg) {
 NORETURN void mp_not_implemented(const char *msg) {
     mp_raise_msg(&mp_type_NotImplementedError, msg);
 }
+
+#if MICROPY_PY_SOFTIRQ
+
+#define MAX_STACK_DEPTH (5)
+
+typedef struct _stack_elem_t {
+    mp_obj_t irqfunc;
+    mp_obj_t arg1;
+    mp_obj_t arg2;
+} stack_elem_t;
+
+volatile bool softirq_enabled = false;
+volatile short sp = -1;
+stack_elem_t stack[MAX_STACK_DEPTH];
+
+void mp_enable_softirq(void) {
+    softirq_enabled = true;
+}
+
+void mp_disable_softirq(void) {
+    mp_uint_t irqstate = disable_irq();
+    softirq_enabled = false;
+    sp = -1;
+    MP_STATE_VM(mp_pending_ex_flags) &= ~PENDING_EX_SOFT_INT;
+    enable_irq(irqstate);
+}
+
+void mp_exec_softirq(void) {
+    stack_elem_t softirq;
+    mp_uint_t irqstate;
+    static bool busy = false;
+
+    if (sp >= 0 && busy == false && __get_IPSR() == 0) {
+        irqstate = disable_irq();
+        softirq = stack[sp--];
+        busy = true;
+        enable_irq(irqstate);
+        if (softirq.arg1 == MP_OBJ_NULL) {
+            mp_call_function_0(softirq.irqfunc);
+        } else if (softirq.arg2 == MP_OBJ_NULL) {
+            mp_call_function_1(softirq.irqfunc, softirq.arg1);
+        } else {
+            mp_call_function_2(softirq.irqfunc, softirq.arg1, softirq.arg2);
+        }
+        irqstate = disable_irq();
+        busy = false;
+        if (sp == -1) {
+            MP_STATE_VM(mp_pending_ex_flags) &= ~PENDING_EX_SOFT_INT;
+        }
+        enable_irq(irqstate);
+    }
+}
+
+mp_obj_t mp_add_softint(mp_obj_t function, mp_obj_t arg1, mp_obj_t arg2) {
+
+    if (sp < MAX_STACK_DEPTH - 1 && softirq_enabled) {
```

Comment:
In my mind a softirq can never behave as a real hard irq, so I don't try to do that. Instead I see a softirq as a kind of exception in the VM. As @pfalcon writes, it is quite easy to get more than one event of the same kind queued. Just the bounces of the user switch is enough to overflow the buffer if the depth is set to two. Regarding the number of arguments stored for each slot, it is actually used for the CAN receive callbacks where the first argument is the CAN object and the second argument is which of three possible reasons for the callback (Message received, receive buffer full and receive buffer overflowed).